interpret: make typed copies lossy wrt provenance and padding

RalfJung opened the pull request:
A "typed copy" in Rust can be a lossy process: when copying at type `usize` (or any other non-pointer type), if the original memory had any provenance, that provenance is lost. When copying at pointer type, if the original memory had partial provenance (i.e., not the same provenance for all bytes), that provenance is lost. When copying any type with padding, the contents of padding are lost.

This PR equips our validity-checking pass with the ability to reset provenance and padding according to those rules. Can be reviewed commit-by-commit. The first three commits are just preparation without any functional change.

Fixes https://github.com/rust-lang/miri/issues/845
Fixes https://github.com/rust-lang/miri/issues/2182

RalfJung committed:
interpret: remove Readable trait, we can use Projectable instead

RalfJung committed:
interpret: make Writeable trait about a to_place operation

RalfJung committed:
interpret: reset provenance on typed copies

RalfJung committed:
interpret: factor out common code for place mutation

RalfJung committed:
interpret: reset padding during validation

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10630197058/job/29468617406) [(plain)](https://github.com/rust-lang/rust/commit/8d243c652e9837ca0788c69c0bb2b1cef6a99e2a/checks/29468617406/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
## Running ui tests in tests/pass for x86_64-unknown-linux-gnu
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-cNuy9h" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" OUT_DIR

FAILED TEST: tests/pass/dyn-star.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-cNuy9h" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/miri_ui/tests/pass" "tests/pass/dyn-star.rs" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/dyn-star.stderr` to the actual output
---
Error: 
   0: ui tests in tests/pass for x86_64-unknown-linux-gnu failed
   1: tests failed

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-6a48dc8439ad88a0 --quiet` (exit status: 1)
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-6a48dc8439ad88a0 --quiet` (exit status: 1)
Command has failed. Rerun with -v to see more details.
  local time: Fri Aug 30 09:55:58 UTC 2024
  network time: Fri, 30 Aug 2024 09:55:58 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10631149991/job/29471467653) [(plain)](https://github.com/rust-lang/rust/commit/c9f962d3071ef73e7ededea8cdf98dd319137c00/checks/29471467653/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
running 2 tests
test test::dev_dependency ... ok
test test::exported_symbol ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered outthread 'rustc' panicked at /checkout/compiler/rustc_const_eval/src/interpret/validity.rs:895:13:
reset_padding on core::num::flt2dec::decoder::FullDecoded: previous field ended at offset 27, next field starts at 26 (and has a size of 1 bytes)
stack backtrace:
   0:     0x7fc4372cf794 - <<std[770d04b741840f10]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[da18e6e8d4edcae8]::fmt::Display>::fmt
   1:     0x7fc43732cee0 - core[da18e6e8d4edcae8]::fmt::write
   2:     0x7fc4372c2ab9 - <std[770d04b741840f10]::sys::pal::unix::stdio::Stderr as std[770d04b741840f10]::io::Write>::write_fmt
   3:     0x7fc4372d2616 - std[770d04b741840f10]::panicking::default_hook::{closure#1}
   4:     0x7fc4372d21f3 - std[770d04b741840f10]::panicking::default_hook
   5:     0x7fc432b09952 - <alloc[46a5234e2585abd6]::boxed::Box<rustc_driver_impl[9ae0f6552e55f858]::install_ice_hook::{closure#0}> as core[da18e6e8d4edcae8]::ops::function::Fn<(&dyn for<'a, 'b> core[da18e6e8d4edcae8]::ops::function::Fn<(&'a std[770d04b741840f10]::panic::PanicHookInfo<'b>,), Output = ()> + core[da18e6e8d4edcae8]::marker::Sync + core[da18e6e8d4edcae8]::marker::Send, &std[770d04b741840f10]::panic::PanicHookInfo)>>::call
   6:     0x7fc4372d31b5 - std[770d04b741840f10]::panicking::rust_panic_with_hook
   7:     0x7fc4372d2d0b - std[770d04b741840f10]::panicking::begin_panic_handler::{closure#0}
   8:     0x7fc4372cfde9 - std[770d04b741840f10]::sys::backtrace::__rust_end_short_backtrace::<std[770d04b741840f10]::panicking::begin_panic_handler::{closure#0}, !>
  10:     0x7fc437328c83 - core[da18e6e8d4edcae8]::panicking::panic_fmt
  10:     0x7fc437328c83 - core[da18e6e8d4edcae8]::panicking::panic_fmt
  11:     0x55fa8ea148a9 - <rustc_const_eval[9bcfc3d414e8b30e]::interpret::validity::ValidityVisitor<miri[29f331f78cefabc5]::machine::MiriMachine>>::reset_padding
  12:     0x55fa8ea76384 - <rustc_const_eval[9bcfc3d414e8b30e]::interpret::eval_context::InterpCx<miri[29f331f78cefabc5]::machine::MiriMachine>>::validate_operand_internal
  13:     0x55fa8ea69185 - <rustc_const_eval[9bcfc3d414e8b30e]::interpret::eval_context::InterpCx<miri[29f331f78cefabc5]::machine::MiriMachine>>::eval_rvalue_into_place
  14:     0x55fa8eaa8329 - <rustc_const_eval[9bcfc3d414e8b30e]::interpret::eval_context::InterpCx<miri[29f331f78cefabc5]::machine::MiriMachine> as miri[29f331f78cefabc5]::concurrency::thread::EvalContextExt>::run_threads
  15:     0x55fa8e9a8561 - miri[29f331f78cefabc5]::eval::eval_entry
  16:     0x55fa8e904b2a - <rustc_middle[51a3c64110132813]::ty::context::GlobalCtxt>::enter::<<miri[2939ea07c9374e44]::MiriCompilerCalls as rustc_driver_impl[9ae0f6552e55f858]::Callbacks>::after_analysis::{closure#0}, ()>
  17:     0x55fa8e8f5f39 - <miri[2939ea07c9374e44]::MiriCompilerCalls as rustc_driver_impl[9ae0f6552e55f858]::Callbacks>::after_analysis
  18:     0x7fc432ab3abb - <rustc_interface[8c616c2973dad9d6]::interface::Compiler>::enter::<rustc_driver_impl[9ae0f6552e55f858]::run_compiler::{closure#0}::{closure#1}, core[da18e6e8d4edcae8]::result::Result<core[da18e6e8d4edcae8]::option::Option<rustc_interface[8c616c2973dad9d6]::queries::Linker>, rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>
  19:     0x7fc432b25135 - rustc_span[1b51c02dd6f2d51c]::create_session_globals_then::<core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>, rustc_interface[8c616c2973dad9d6]::util::run_in_thread_with_globals<rustc_interface[8c616c2973dad9d6]::util::run_in_thread_pool_with_globals<rustc_interface[8c616c2973dad9d6]::interface::run_compiler<core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>, rustc_driver_impl[9ae0f6552e55f858]::run_compiler::{closure#0}>::{closure#1}, core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>::{closure#0}, core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
  20:     0x7fc432a893a2 - std[770d04b741840f10]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[8c616c2973dad9d6]::util::run_in_thread_with_globals<rustc_interface[8c616c2973dad9d6]::util::run_in_thread_pool_with_globals<rustc_interface[8c616c2973dad9d6]::interface::run_compiler<core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>, rustc_driver_impl[9ae0f6552e55f858]::run_compiler::{closure#0}>::{closure#1}, core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>::{closure#0}, core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>
  21:     0x7fc432b6454f - <<std[770d04b741840f10]::thread::Builder>::spawn_unchecked_<rustc_interface[8c616c2973dad9d6]::util::run_in_thread_with_globals<rustc_interface[8c616c2973dad9d6]::util::run_in_thread_pool_with_globals<rustc_interface[8c616c2973dad9d6]::interface::run_compiler<core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>, rustc_driver_impl[9ae0f6552e55f858]::run_compiler::{closure#0}>::{closure#1}, core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>::{closure#0}, core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[da18e6e8d4edcae8]::result::Result<(), rustc_span[1b51c02dd6f2d51c]::ErrorGuaranteed>>::{closure#1} as core[da18e6e8d4edcae8]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  22:     0x7fc4372df385 - <std[770d04b741840f10]::sys::pal::unix::thread::Thread>::new::thread_start
  24:     0x7fc430ae2850 - <unknown>
  25:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.
error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/miri/issues/new

note: please make sure that you have updated to the latest nightly

warning: the ICE couldn't be written to `/checkout/src/tools/miri/test-cargo-miri/rustc-ice-2024-08-30T11_05_20-88427.txt`: Read-only file system (os error 30)
note: rustc 1.82.0-nightly (ad4fdf979 2024-08-30) running on x86_64-unknown-linux-gnu


note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C split-debuginfo=off -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -Z unstable-options -Z on-broken-pipe=kill -Z binary-dep-depinfo
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
end of query stack

Miri caused an ICE during evaluation. Here's the interpreter backtrace at the time of the panic:
note: the place in the program where the ICE was triggered
  --> /checkout/library/core/src/num/flt2dec/decoder.rs:99:16
   |
99 |     (sign < 0, decoded)
   |
   = note: BACKTRACE:
   = note: BACKTRACE:
   = note: inside `core::num::flt2dec::decoder::decode::<f64>` at /checkout/library/core/src/num/flt2dec/decoder.rs:99:16: 99:23
   = note: inside `core::num::flt2dec::to_exact_fixed_str::<'_, f64, for<'b, 'a> fn(&'b core::num::flt2dec::decoder::Decoded, &'a mut [std::mem::MaybeUninit<u8>], i16) -> (&'a [u8], i16) {core::num::flt2dec::strategy::grisu::format_exact}>` at /checkout/library/core/src/num/flt2dec/mod.rs:604:36: 604:45
   = note: inside `core::fmt::float::float_to_decimal_common_exact::<f64>` at /checkout/library/core/src/fmt/float.rs:39:21: 46:6
   = note: inside `core::fmt::float::float_to_decimal_display::<f64>` at /checkout/library/core/src/fmt/float.rs:90:9: 90:65
   = note: inside `core::fmt::float::<impl std::fmt::Display for f64>::fmt` at /checkout/library/core/src/fmt/float.rs:210:17: 210:52
   = note: inside `core::fmt::rt::Argument::<'_>::fmt` at /checkout/library/core/src/fmt/rt.rs:177:76: 177:95
   = note: inside `core::fmt::run` at /checkout/library/core/src/fmt/mod.rs:1228:14: 1228:28
   = note: inside `std::fmt::Formatter::<'_>::write_fmt` at /checkout/library/core/src/fmt/mod.rs:1658:13: 1658:33
   = note: inside `<test::test::time::TestSuiteExecTime as std::fmt::Display>::fmt` at /checkout/library/core/src/macros/mod.rs:650:9: 650:55
   = note: inside `<test::test::time::TestSuiteExecTime as std::fmt::Display>::fmt` at /checkout/library/core/src/macros/mod.rs:650:9: 650:55
   = note: inside `<&test::test::time::TestSuiteExecTime as std::fmt::Display>::fmt` at /checkout/library/core/src/fmt/mod.rs:2382:62: 2382:82
   = note: inside `core::fmt::rt::Argument::<'_>::fmt` at /checkout/library/core/src/fmt/rt.rs:177:76: 177:95
   = note: inside `<&mut std::string::String as std::fmt::Write::write_fmt::SpecWriteFmt>::spec_write_fmt` at /checkout/library/core/src/fmt/mod.rs:226:21: 226:38
   = note: inside `<std::string::String as std::fmt::Write>::write_fmt` at /checkout/library/core/src/fmt/mod.rs:231:9: 231:34
   = note: inside `std::fmt::format::format_inner` at /checkout/library/alloc/src/fmt.rs:636:9: 637:29
   = note: inside closure at /checkout/library/alloc/src/fmt.rs:642:34: 642:52
   = note: inside closure at /checkout/library/alloc/src/fmt.rs:642:34: 642:52
   = note: inside `std::option::Option::<&str>::map_or_else::<std::string::String, {closure@std::fmt::format::{closure#0}}, for<'a> fn(&'a str) -> <str as std::borrow::ToOwned>::Owned {<str as std::borrow::ToOwned>::to_owned}>` at /checkout/library/core/src/option.rs:1211:21: 1211:30
   = note: inside `std::fmt::format` at /checkout/library/alloc/src/fmt.rs:642:5: 642:87
   = note: inside `<test::test::formatters::pretty::PrettyFormatter<std::io::Stdout> as test::test::formatters::OutputFormatter>::write_run_finish` at /checkout/library/alloc/src/macros.rs:128:23: 128:84
   = note: inside `test::test::run_tests_console` at /checkout/library/test/src/console.rs:327:5: 327:30
   = note: inside `test::test::test_main_static` at /checkout/library/test/src/lib.rs:159:5: 159:40
   = note: inside `main`
   = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /checkout/library/core/src/ops/function.rs:250:5: 250:71
   = note: inside `std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /checkout/library/std/src/sys/backtrace.rs:154:18: 154:21
   = note: inside `std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /checkout/library/std/src/sys/backtrace.rs:154:18: 154:21
   = note: inside closure at /checkout/library/std/src/rt.rs:164:18: 164:75
   = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /checkout/library/core/src/ops/function.rs:284:13: 284:31
   = note: inside `std::panicking::r#try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panicking.rs:554:40: 554:43
   = note: inside `std::panicking::r#try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /checkout/library/std/src/panicking.rs:518:19: 518:88
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panic.rs:345:14: 345:33
   = note: inside closure at /checkout/library/std/src/rt.rs:143:48: 143:73
   = note: inside `std::panicking::r#try::do_call::<{closure@std::rt::lang_start_internal::{closure#2}}, isize>` at /checkout/library/std/src/panicking.rs:554:40: 554:43
   = note: inside `std::panicking::r#try::<isize, {closure@std::rt::lang_start_internal::{closure#2}}>` at /checkout/library/std/src/panicking.rs:518:19: 518:88
   = note: inside `std::panic::catch_unwind::<{closure@std::rt::lang_start_internal::{closure#2}}, isize>` at /checkout/library/std/src/panic.rs:345:14: 345:33
   = note: inside `std::rt::lang_start::<()>` at /checkout/library/std/src/rt.rs:163:17: 168:6

error: test failed, to rerun pass `--bin cargo-miri-test`
error: test failed, to rerun pass `--bin cargo-miri-test`
Command has failed. Rerun with -v to see more details.
  local time: Fri Aug 30 11:05:21 UTC 2024
  network time: Fri, 30 Aug 2024 11:05:22 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10632362902/job/29475411157) [(plain)](https://github.com/rust-lang/rust/commit/cd7630578c5f326c4f81b32a643c2fb507af0f9e/checks/29475411157/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

## Running ui tests in tests/fail for i686-pc-windows-msvc
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-ikbihp" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" OUT_DIR

FAILED TEST: tests/fail/provenance/ptr_copy_loses_partial_provenance0.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-ikbihp" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/miri_ui/tests/fail/provenance" "tests/fail/provenance/ptr_copy_loses_partial_provenance0.rs" "--edition" "2021"
error: fail test got exit status: 0, but expected 1

error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/fail/provenance/ptr_copy_loses_partial_provenance0.stderr` to the actual output
--- tests/fail/provenance/ptr_copy_loses_partial_provenance0.stderr
+++ <stderr output>
-error: Undefined Behavior: memory access failed: expected a pointer to 4 bytes of memory, but got $HEX[noalloc] which is a dangling pointer (it has no provenance)
-  --> $DIR/ptr_copy_loses_partial_provenance0.rs:LL:CC
-LL |         let _val = *ptr;
-LL |         let _val = *ptr;
-   |                    ^^^^ memory access failed: expected a pointer to 4 bytes of memory, but got $HEX[noalloc] which is a dangling pointer (it has no provenance)
-   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
-   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
-   = note: BACKTRACE:
-   = note: inside `main` at $DIR/ptr_copy_loses_partial_provenance0.rs:LL:CC
---

error: `dangling` not found in diagnostics on line 15
##[error]  --> tests/fail/provenance/ptr_copy_loses_partial_provenance0.rs:15:36
   |
15 |         let _val = *ptr; //~ERROR: dangling
   |

full stderr:


full stdout:



FAILED TEST: tests/fail/provenance/ptr_copy_loses_partial_provenance1.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-ikbihp" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/miri_ui/tests/fail/provenance" "tests/fail/provenance/ptr_copy_loses_partial_provenance1.rs" "--edition" "2021"
error: fail test got exit status: 0, but expected 1

error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/fail/provenance/ptr_copy_loses_partial_provenance1.stderr` to the actual output
--- tests/fail/provenance/ptr_copy_loses_partial_provenance1.stderr
+++ <stderr output>
-error: Undefined Behavior: memory access failed: expected a pointer to 4 bytes of memory, but got $HEX[noalloc] which is a dangling pointer (it has no provenance)
-  --> $DIR/ptr_copy_loses_partial_provenance1.rs:LL:CC
-LL |         let _val = *ptr;
-LL |         let _val = *ptr;
-   |                    ^^^^ memory access failed: expected a pointer to 4 bytes of memory, but got $HEX[noalloc] which is a dangling pointer (it has no provenance)
-   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
-   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
-   = note: BACKTRACE:
-   = note: inside `main` at $DIR/ptr_copy_loses_partial_provenance1.rs:LL:CC
---

error: `dangling` not found in diagnostics on line 15
##[error]  --> tests/fail/provenance/ptr_copy_loses_partial_provenance1.rs:15:36
   |
15 |         let _val = *ptr; //~ERROR: dangling
   |

full stderr:

---
Error: 
   0: ui tests in tests/fail for i686-pc-windows-msvc failed
   1: tests failed

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-6a48dc8439ad88a0 --quiet` (exit status: 1)
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-6a48dc8439ad88a0 --quiet` (exit status: 1)
Command has failed. Rerun with -v to see more details.
  local time: Fri Aug 30 12:49:19 UTC 2024
  network time: Fri, 30 Aug 2024 12:49:20 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

michaelwoerister commented:
r? compiler

saethlin commented:
r? saethlin

RalfJung commented on compiler/rustc_const_eval/src/interpret/validity.rs:
```diff
@@ -204,11 +206,58 @@ fn write_path(out: &mut String, path: &[PathElem]) {
+// These are (offset, length) pairs, and they are sorted and mutually disjoint,
+// and never adjacent (i.e. there's always a gap between two of them).
+#[derive(Debug, Clone)]
+pub struct RangeSet(Vec<(Size, Size)>);
```
This is kind of like `IntervalSet` in `rustc_index`, except we need u64 and not just u32.  (Also `IntervalSet` has a `domain` which we don't need, but we could just set `u64::MAX`.)

Alternatively, https://docs.rs/rangemap/1.5.1/rangemap/set/struct.RangeSet.html would probably also do it. But that would add a new dependency.

RalfJung replied:
Maybe I should try making `IntervalSet` generic over a `T: Into<u64>`? That will make it less efficient (computation-wise, not storage-wise) on a 32bit system. Not sure if we care.

The alternative to `Into<u64>` is to basically require the equivalent of a `Num` trait, which might get awkward, but I haven't tried.

RalfJung replied:
Ah no `Into<u64>` is not enough. We need at least things like `max`, though that is easy via `PartialOrd`. More problematic is the fact that IntervalSet uses inclusive ranges for some reason, so it has to do a lot of annoying `+1`. My implementation uses right-exclusive ranges which, as usual, makes the arithmetic a lot nicer -- but maybe someone benchmarked this and  anyway I'm not going to re-do IntervalSet now.^^

Cc @Mark-Simulacrum since you seem to have added `IntervalSet` a couple of years ago.

saethlin replied:
FWIW I'm perfectly happy to see new data structures land in random places initially with highly specific implementations. If someone finds a use for it somewhere else, that's the point at which it becomes worth it to make things reusable.

saethlin commented on compiler/rustc_const_eval/src/const_eval/machine.rs:
```diff
@@ -65,6 +65,9 @@ pub struct CompileTimeMachine<'tcx> {
     /// Used to prevent reads from a static's base allocation, as that may allow for self-initialization loops.
     pub(crate) static_root_ids: Option<(AllocId, LocalDefId)>,
+
+    /// A cache of "data range" computations for unions (i.e., the offsets of non-padding bytes).
```
This seems like the sort of thing that would normally be a query. Is there a reason it isn't?

RalfJung replied:
I'm not sure we even want this cached during a normal compilation (but it probably won't hurt much either), and the `RangeSet` type is currently only declared in the interpreter so can't be used as a query. Also I am largely unfamiliar with the query system so adding a query is scary. ;)

RalfJung replied:
If we have many constants of the same (union) type, then we would encounter it repeatedly and caching might be worth it. But we don't do validation during const-eval by default so the cache is not nearly as necessary as for Miri.

I can see how making this a query makes sense. But it would probably be a query we wouldn't want to serialize for incremental (I assume there's a way to do that), and as I said I am not sure where we would then put the `RangeSet`. Probably `rustc_data_structures`, but that seems odd for such a specific data structure. I'm open for suggestions though. :)

saethlin replied:
You've made a good case that we shouldn't make this into a query in this PR, I'm filing away this conversation for the next time I'm squinting at interpreter profiles :)

saethlin commented on compiler/rustc_const_eval/src/util/check_validity_requirement.rs:
```diff
@@ -61,13 +61,17 @@ fn might_permit_raw_init_strict<'tcx>(
+        .validate_operand(
+            &allocated.into(),
+            /*recursive*/ false,
+            /*reset_provenance_and_padding*/ false,
```
Why does this single caller not want provenance and padding reset? I looked at the docs for this function for context and I think they refer to a function `might_permit_raw_init` that doesn't exist?

RalfJung replied:
This is used for assert_zero_valid / assert_mem_uninitialized_valid. The CompileTimeMachine that we operate in here is discarded at the end of this function, alongside the data we are validating here. So there's no reason to reset its padding and provenance, nobody will ever see those writes anyway.

saethlin replied:
FYI I have taken to always using the hyperlink rustdoc syntax when mentioning other functions so that the link checker ensures that I can't mention a function that doesn't exist. The habit has already prevented a few of my doc comments from rotting. It's extra work, up to you if you want to adopt it.

RalfJung replied:
> So this function parameter and all the branching on it is supposed to be a performance optimization?

Yes.

Also this is not the only caller that doesn't to the reset -- `const_validate_operand` also doesn't set it. Basically I wanted this change to be Miri-only.

And `validate_operand` in the recursive case, when it validates behind references, also doesn't reset padding and provenance -- and that is crucial, it would be wrong to reset this (and indeed the Miri test suite fails when you try, since if you ever have a reference to read-only memory it then blows up when trying to reset padding and provenance there).

saethlin commented:
I'm very happy to see less interior mutability. The compiler uses more casual Cell and RefCell than any other codebase I've worked in which seems ironic.

@rustbot author

RalfJung committed:
clarify comments and names in check_validity_requirement

RalfJung committed:
IntervalSet: add comment about representation

saethlin commented:
@rustbot review

bors commented:
:umbrella: The latest upstream changes (presumably #129346) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

saethlin commented:
Gah I was beaten to this by a conflict.

r=me once that's fixed

RalfJung commented:
@bors r=saethlin rollup=iffy

bors commented:
:pushpin: Commit 11bd99de8cc67683f215317dba55c91aaf3b5767 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 11bd99de8cc67683f215317dba55c91aaf3b5767 -->
<!-- homu: {"type":"Approved","sha":"11bd99de8cc67683f215317dba55c91aaf3b5767","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 11bd99de8cc67683f215317dba55c91aaf3b5767 with merge 175c0f33b841a6861494d9541a7cbf0a75d6ae39...
<!-- homu: {"type":"BuildStarted","head_sha":"11bd99de8cc67683f215317dba55c91aaf3b5767","merge_sha":"175c0f33b841a6861494d9541a7cbf0a75d6ae39"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10761668069/job/29841138717) [(plain)](https://github.com/rust-lang-ci/rust/commit/175c0f33b841a6861494d9541a7cbf0a75d6ae39/checks/29841138717/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10761668069/job/29844272719)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10761668069/job/29844272719","builder_name":"checks-actions"} -->

RalfJung commented:
Oh... does this make the miri-test-libstd part so slow that we hit a timeout...

RalfJung commented:
AFAIK this uses a Miri built with debug assertions. That's probably just overkill.

RalfJung commented:
Oh, we are already setting `NO_DEBUG_ASSERTIONS=1`... does this really have such a huge perf impact? We're already doing tons of little reads for validation, I would not have thought that a few writes make such a big difference.

RalfJung commented:
Checking this with the Miri benchmark suite (and comparing against a Miri built against a release rustc, which may be unfair as that's build with a bunch of fancy things like PGO that my local build does not get), running with `-Zmiri-disable-stacked-borrows` to make validation matter more:
- backtraces: +6%
- big-allocs: +14%
- mse: +21%
- slice-get-unchecked: +21%

The others are in the same ballpark, I just don't want to manually do the math for all of them. ;)

So... I don't think that can explain why the aux builder went from 1h40min to >4h. I'll do a try build to check if the slowdown is reproducible.

RalfJung commented:
@bors try

bors commented:
:hourglass: Trying commit 11bd99de8cc67683f215317dba55c91aaf3b5767 with merge 4d8ac937c137092a594c16af563195fdf1d117d1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"11bd99de8cc67683f215317dba55c91aaf3b5767","merge_sha":"4d8ac937c137092a594c16af563195fdf1d117d1"} -->

RalfJung commented:
:thinking: 
```
2024-09-08T18:07:47.0014842Z test iter::adapters::flatten::test_iterator_flatten_fold ... ok
2024-09-08T21:08:22.3047668Z ##[error]The operation was canceled.
```
So whatever test comes after `test_iterator_flatten_fold` just took 3h...

Usually this looks like
```
2024-09-09T00:39:46.2237911Z test iter::adapters::flatten::test_iterator_flatten_fold ... ok
2024-09-09T00:39:46.3841795Z test iter::adapters::flatten::test_trusted_len_flatten ... ok
```

And indeed I can reproduce this locally with `./x.py miri test library/core -- iter`.

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10768959821/job/29859150519) [(plain)](https://github.com/rust-lang-ci/rust/commit/4d8ac937c137092a594c16af563195fdf1d117d1/checks/29859150519/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10768959821/job/29870810369)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10768959821/job/29870810369","builder_name":"checks-actions"} -->

RalfJung committed:
union padding computation: add fast-path for ZST

Also avoid even tracking empty ranges, and add fast-path for arrays of scalars

RalfJung commented:
Indeed, the problem is a union with a field of type `[(); usize::MAX]`. The last commit should fix that.

RalfJung commented:
@rustbot ready

RalfJung commented:
@bors try
Let's see how long this takes now.

bors commented:
:hourglass: Trying commit 65c70900ceda1c276c8eef1dad4e2c4f7a0756d0 with merge 7e2aa1784cff4b926ea42fb4e8b5118b4a4b825e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"65c70900ceda1c276c8eef1dad4e2c4f7a0756d0","merge_sha":"7e2aa1784cff4b926ea42fb4e8b5118b4a4b825e"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10773532615/job/29873542737) [(plain)](https://github.com/rust-lang-ci/rust/commit/7e2aa1784cff4b926ea42fb4e8b5118b4a4b825e/checks/29873542737/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
test mem::const_maybe_uninit ... ok
error: Undefined Behavior: constructing invalid value at [8]: encountered uninitialized memory, but expected an integer
##[error]    --> core/tests/mem.rs:791:25
     |
791  |     assert_eq!(unsafe { (*UNINIT.0.cast::<[[u8; SIZE]; 1]>())[0] }, [0u8; SIZE]);
     |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value at [8]: encountered uninitialized memory, but expected an integer
     = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
     = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
     = note: BACKTRACE on thread `mem::const_mayb`:
     = note: inside `mem::const_maybe_uninit_zeroed` at core/tests/mem.rs:791:25: 791:65
---
     |                               ^
     |
    ::: /checkout/library/core/src/macros/mod.rs:1659:5
     |
1659 |     pub macro test($item:item) {
     |     -------------- in this expansion of `#[test]`
note: inside `<{closure@core/tests/mem.rs:774:1: 792:2} as std::ops::FnOnce<()>>::call_once - shim`
     |
     |
250  |     extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
note: inside `<fn() -> std::result::Result<(), std::string::String> as std::ops::FnOnce<()>>::call_once - shim(fn() -> std::result::Result<(), std::string::String>)`
    --> /checkout/library/core/src/ops/function.rs:250:5
     |
     |
250  |     extern "rust-call" fn call_once(self, args: Args) -> Self::Output;

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
error: aborting due to 1 previous error

error: test failed, to rerun pass `-p core --test coretests`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/cargo-miri runner /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/miri/x86_64-unknown-linux-gnu/debug/deps/coretests-40e39f8fd5de5322 -Z unstable-options --format json` (exit status: 1)
note: test exited abnormally; to see the full output pass --nocapture to the harness.
Build completed unsuccessfully in 0:14:36
make: *** [Makefile:53: check-aux] Error 1
  network time: Mon, 09 Sep 2024 13:53:01 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10773532615/job/29877245655)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10773532615/job/29877245655","builder_name":"checks-actions"} -->

RalfJung committed:
fix UB in a test

also add an explicit test for the fact that a Option<WidePtr> has padding when it is None

RalfJung commented:
Well, it did not take 4h this time...

... instead it found actual UB in a test.

RalfJung commented:
@bors try

bors commented:
:hourglass: Trying commit 0a70924c2160c8f93a446328e207a1983e62596a with merge ecd72e47bb2b451038ec8ce04f1fc09a588c3416...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0a70924c2160c8f93a446328e207a1983e62596a","merge_sha":"ecd72e47bb2b451038ec8ce04f1fc09a588c3416"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10775133568/job/29884683219)
Build commit: ecd72e47bb2b451038ec8ce04f1fc09a588c3416 (`ecd72e47bb2b451038ec8ce04f1fc09a588c3416`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10775133568/job/29884683219"},"merge_sha":"ecd72e47bb2b451038ec8ce04f1fc09a588c3416"} -->

RalfJung commented:
Much better :)

RalfJung commented:
@rustbot ready

saethlin commented:
The fix looks good to me! 
@bors r+

bors commented:
:pushpin: Commit 0a70924c2160c8f93a446328e207a1983e62596a has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 0a70924c2160c8f93a446328e207a1983e62596a -->
<!-- homu: {"type":"Approved","sha":"0a70924c2160c8f93a446328e207a1983e62596a","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 0a70924c2160c8f93a446328e207a1983e62596a with merge ecb46242b1aeb783ac13f94644c05d85316653f6...
<!-- homu: {"type":"BuildStarted","head_sha":"0a70924c2160c8f93a446328e207a1983e62596a","merge_sha":"ecb46242b1aeb783ac13f94644c05d85316653f6"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc-ext2`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10782229461/job/29901827870) [(plain)](https://github.com/rust-lang-ci/rust/commit/ecb46242b1aeb783ac13f94644c05d85316653f6/checks/29901827870/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] miri test:false 4.488
error: failed to remove file `C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage1-tools\x86_64-pc-windows-msvc\release\miri.exe`

Caused by:
  Access is denied. (os error 5)
Command has failed. Rerun with -v to see more details.
  local time: Mon, Sep  9, 2024 11:14:58 PM
  network time: Mon, 09 Sep 2024 23:14:58 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10782229461/job/29903377497)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10782229461/job/29903377497","builder_name":"checks-actions"} -->

saethlin commented:
@bors retry https://github.com/rust-lang/rust/issues/127883

bors commented:
:hourglass: Testing commit 0a70924c2160c8f93a446328e207a1983e62596a with merge 304b7f801bab31233680879ca4fb6eb294706a59...
<!-- homu: {"type":"BuildStarted","head_sha":"0a70924c2160c8f93a446328e207a1983e62596a","merge_sha":"304b7f801bab31233680879ca4fb6eb294706a59"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10784583238/job/29911890142)
Approved by: saethlin
Pushing 304b7f801bab31233680879ca4fb6eb294706a59 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"saethlin","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10784583238/job/29911890142"},"merge_sha":"304b7f801bab31233680879ca4fb6eb294706a59"} -->

rust-timer commented:
Finished benchmarking commit (304b7f801bab31233680879ca4fb6eb294706a59): [comparison URL](https://perf.rust-lang.org/compare.html?start=712463de61c65033a6f333f0a14fbb65e34efc50&end=304b7f801bab31233680879ca4fb6eb294706a59&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=712463de61c65033a6f333f0a14fbb65e34efc50&end=304b7f801bab31233680879ca4fb6eb294706a59&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.1% | [0.2%, 4.7%] | 13    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=712463de61c65033a6f333f0a14fbb65e34efc50&end=304b7f801bab31233680879ca4fb6eb294706a59&stat=max-rss)
<details>
<summary>Results (secondary -2.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-3.7%, -1.4%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=712463de61c65033a6f333f0a14fbb65e34efc50&end=304b7f801bab31233680879ca4fb6eb294706a59&stat=cycles:u)
<details>
<summary>Results (secondary -2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.3%, -2.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=712463de61c65033a6f333f0a14fbb65e34efc50&end=304b7f801bab31233680879ca4fb6eb294706a59&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 757.734s -> 756.628s (-0.15%)
**Artifact size**: 341.11 MiB -> 341.23 MiB (0.03%)

RalfJung commented:
Oh wow, just having these branches sit there, even if they are never taken, shows up so strongly in the benchmarks?

I could add a `Machine`  associated const that should make the branches trivially optimized away... but that will also mean one cannot enable resetting padding with `-Zextra-const-ub-checks` any more.

Kobzol commented:
It only got a significant regression on [CTFE stress tests](https://github.com/rust-lang/rustc-perf/blob/master/collector/compile-benchmarks/ctfe-stress-5/src/lib.rs). If you don't think that this should affect real programs, I think it would be fine to just take the hit (as it sounds that the proposed optimization would introduce a lot of complexity into the code).

RalfJung commented:
It wouldn't be a lot of complexity, but it would make the (nightly-only) const-eval UB check less effective. But those checks already miss a bunch of UB anyway so maybe that's okay.

djkoloski commented:
Is it intended that after this change `MaybeUninit::zeroed()` no longer returns a completely-zeroed value? Because it's being returned from a function, the padding bytes of the interior `T` are being uninitialized.

I ran into this just now in rkyv, as I was using `MaybeUninit::zeroed()` to make a zeroed struct that eventually gets cast to bytes and written to a vector.

saethlin commented:
Function returns are a typed copy. This PR only changed the behavior of the const eval interpreter, codegen has always been capable of breaking such code.

RalfJung commented:
Yes, this is intended -- padding bytes are always initialized.

Happy to hear that is already found a bug in the wild. :D Is there something we can link to for this bug? If you wouldn't mind I'd like to add this to the "bugs Miri found" list.

djkoloski commented:
> Function returns are a typed copy.

Of course. I think the more salient question here is "what does a typed copy mean for unions?". Is there any documentation about what bytes are considered padding in unions? Is just `MaybeUninit` special? I think there's an argument to be made that unions having padding doesn't make sense, as what's stored in a union is not generally known to the compiler. I personally think "unions have padding where every field has padding" could make sense too, but I haven't seen any documentation about this as far as I can remember.

> Is there something we can link to for this bug?

I think the best I have is this [failed CI run](https://github.com/rkyv/rkyv/actions/runs/10803927210/job/29968560635). I haven't checked for other potentially-affected sources yet. Feel free to add whatever you'd like! 🙂

RalfJung commented:
> I think the more salient question here is "what does a typed copy mean for unions?". 

Yes, that is indeed non-trivial.

I would love it to be a fully copy of all bytes, but sadly, [that does not work](https://github.com/rust-lang/unsafe-code-guidelines/issues/156). So currently we use "a union has padding where all fields have padding". I'd love to change this (see https://github.com/rust-lang/unsafe-code-guidelines/issues/518), but I am not sure if that is possible.
